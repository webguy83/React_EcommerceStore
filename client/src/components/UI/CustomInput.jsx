import React from "react";
import styled, { css } from "styled-components";

// css

export const shrinkText = css`
  top: -1rem;
  font-size: 1.3rem;
  transition: all 0.3s ease-out;
`;

export const normalText = css`
  top: -0.5rem;
  font-size: 1.6rem;
  transition: all 0.3s ease-in;
`;

const CustomLabel = styled.label`
  color: darkgray;
  position: absolute;

  ${normalText}

  ${({ ...otherInputProps }) => {
    if (otherInputProps.value.length > 0) {
      return shrinkText;
    }
  }}
`;
const Input = styled.input`
  font-family: inherit;
  border: 1px solid darkgray;

  &:focus {
    outline: none;

    & + ${CustomLabel} {
      ${shrinkText}
    }
  }

  &[type="password"],
  &[type="textarea"] {
    letter-spacing: 0.3rem;
  }
`;

const TextArea = styled.textarea`
  font-family: inherit;
  &:focus {
    outline: none;

    & + ${CustomLabel} {
      ${shrinkText}
    }
  }
`;

const ElementGroup = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  padding: 1.5rem 0;
`;

const CustomInput = ({
  label,
  labelToInputLink,
  handleChange,
  ...otherInputProps
}) => {
  const { type } = otherInputProps;
  return (
    <ElementGroup>
      {type === "textarea" ? (
        <TextArea
          data-testid="textarea"
          id={labelToInputLink}
          onChange={handleChange}
          rows="5"
          {...otherInputProps}
        />
      ) : (
        <Input
          data-testid="input"
          id={labelToInputLink}
          onChange={handleChange}
          {...otherInputProps}
        />
      )}

      {label ? (
        <CustomLabel
          data-testid="label"
          htmlFor={labelToInputLink}
          {...otherInputProps}
        >
          {label}
        </CustomLabel>
      ) : null}
    </ElementGroup>
  );
};

export default CustomInput;
