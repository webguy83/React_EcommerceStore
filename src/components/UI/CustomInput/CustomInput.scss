@mixin shrinkText {
    top: -2rem;
    font-size: 1.3rem;
    transition: all .3s ease-out;
}

.customInput {
    border: none;
    border-bottom: 1px solid black;

    &:focus {
        outline: none;

        &+.customLabel {
            @include shrinkText();
        }
    }

    &[type=password] {
        letter-spacing: .3rem;
    }
}

.customLabel {
    font-size: 1.6rem;
    color: darkgray;
    position: absolute;
    top: -0.5rem;
    transition: all .3s ease-in;

    &.shrink  {
        @include shrinkText();
    }
}

.elementGroup {
    display: flex;
    flex-direction: column;
    margin-top: 3rem;
    position: relative;
    padding: 1rem 0;
}