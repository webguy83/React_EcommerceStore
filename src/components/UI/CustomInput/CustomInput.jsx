import React from 'react';

import './CustomInput.scss';

export default ({ label, handleChange, ...otherInputProps }) => {
    return (
        <div className="elementGroup">
            <input id={label ? label.toLowerCase() : null} className="customInput" onChange={handleChange} {...otherInputProps} />
            {
                label ?
                    <label htmlFor={label ? label.toLowerCase() : null} className={`${otherInputProps.value.length ? 'shrink' : ''} customLabel`}>{label}</label>
                    :
                    null
            }
            
        </div>
    );
};