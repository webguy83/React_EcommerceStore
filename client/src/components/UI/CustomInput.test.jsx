import CustomInput, { shrinkText, normalText } from "./CustomInput";
import React from "react";
import "jest-styled-components";
import "@testing-library/jest-dom/extend-expect";
import { render, cleanup } from "@testing-library/react";

describe("default tests for Custom Input", () => {
  afterEach(cleanup);

  it("should render without crashing and input as default", () => {
    const { asFragment, getByTestId } = render(
      <CustomInput
        type="text"
        name="firstName"
        handleChange={() => {}}
        value="Curtis"
        label="First Name"
        labelToInputLink="firstName"
        required
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("input").value).toBe("Curtis");
  });

  it("should include a label element if label is defined", () => {
    const { getByTestId } = render(
      <CustomInput
        type="text"
        name="firstName"
        handleChange={() => {}}
        value="Curtis"
        label="First Name"
        labelToInputLink="firstName"
        required
      />
    );
    expect(getByTestId("label")).toHaveTextContent("First Name");
  });

  it("should not include a label element if label is not defined", () => {
    const { queryByTestId } = render(
      <CustomInput
        type="text"
        name="firstName"
        handleChange={() => {}}
        value="Bob"
        required
      />
    );
    expect(queryByTestId("label")).toBeNull();
  });

  it("should render a textarea", () => {
    const { getByTestId } = render(
      <CustomInput
        type="textarea"
        name="comments"
        label="comments"
        labelToInputLink="comments"
        handleChange={() => {}}
        value="Bob is fun"
        required
      />
    );
    expect(getByTestId("textarea")).toHaveTextContent("Bob is fun");
  });
});

describe("CSS testing", () => {
  afterEach(cleanup);

  it("should shrink label text if text in value field", () => {
    const { getByTestId } = render(
      <CustomInput
        type="text"
        name="firstName"
        handleChange={() => {}}
        value="Curtis"
        label="First Name"
        labelToInputLink="firstName"
        required
      />
    );
    expect(getByTestId("label")).toHaveStyle(shrinkText[0]);
  });

  it("should not shrink label text if no text in value field", () => {
    const { getByTestId } = render(
      <CustomInput
        type="text"
        name="firstName"
        handleChange={() => {}}
        value=""
        label="First Name"
        labelToInputLink="firstName"
        required
      />
    );
    expect(getByTestId("label")).toHaveStyle(normalText[0]);
  });
});
