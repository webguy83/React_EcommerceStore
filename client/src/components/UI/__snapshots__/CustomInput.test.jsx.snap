// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default tests for Custom Input should render without crashing and input as default 1`] = `
<DocumentFragment>
  .c3 {
  color: darkgray;
  position: absolute;
  top: -0.5rem;
  font-size: 1.6rem;
  -webkit-transition: all 0.3s ease-in;
  transition: all 0.3s ease-in;
  top: -1rem;
  font-size: 1.3rem;
  -webkit-transition: all 0.3s ease-out;
  transition: all 0.3s ease-out;
}

.c1 {
  font-family: inherit;
  border: 1px solid darkgray;
}

.c1:focus {
  outline: none;
}

.c1:focus + .c2 {
  top: -1rem;
  font-size: 1.3rem;
  -webkit-transition: all 0.3s ease-out;
  transition: all 0.3s ease-out;
}

.c1[type="password"],
.c1[type="textarea"] {
  -webkit-letter-spacing: 0.3rem;
  -moz-letter-spacing: 0.3rem;
  -ms-letter-spacing: 0.3rem;
  letter-spacing: 0.3rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding: 1.5rem 0;
}

<div
    class="c0"
  >
    <input
      class="c1"
      data-testid="input"
      id="firstName"
      name="firstName"
      required=""
      type="text"
      value="Curtis"
    />
    <label
      class="c2 c3"
      data-testid="label"
      for="firstName"
      name="firstName"
      required=""
      type="text"
      value="Curtis"
    >
      First Name
    </label>
  </div>
</DocumentFragment>
`;
